File: imagepipeline/src/main/java/com/facebook/imagepipeline/transcoder/SimpleImageTranscoder.java
/*
 * Copyright (c) 2018-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
package com.facebook.imagepipeline.transcoder;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Matrix;
import android.os.Build;
import com.facebook.common.logging.FLog;
import com.facebook.imageformat.DefaultImageFormats;
import com.facebook.imageformat.ImageFormat;
import com.facebook.imagepipeline.common.ResizeOptions;
import com.facebook.imagepipeline.common.RotationOptions;
import com.facebook.imagepipeline.image.EncodedImage;
import java.io.OutputStream;
import javax.annotation.Nullable;

/**
 * Image transcoder using only the Android API. Clients can use this if they don't want to use the
 * native implementation. This image transcoder requires more memory.
 */
public class SimpleImageTranscoder implements ImageTranscoder {
  private static final String TAG = "SimpleImageTranscoder";
  private final boolean mResizingEnabled;
  private final int mMaxBitmapSize;

  public SimpleImageTranscoder(final boolean resizingEnabled, final int maxBitmapSize) {
    mResizingEnabled = resizingEnabled;
    mMaxBitmapSize = maxBitmapSize;
  }

  @Override
  public ImageTranscodeResult transcode(
      EncodedImage encodedImage,
      OutputStream outputStream,
      @Nullable RotationOptions rotationOptions,
      @Nullable ResizeOptions resizeOptions,
      @Nullable ImageFormat outputFormat,
      @Nullable Integer quality) {
    if (quality == null) {
      quality = JpegTranscoderUtils.DEFAULT_JPEG_QUALITY;
    }
    if (rotationOptions == null) {
      rotationOptions = RotationOptions.autoRotate();
    }

    final int sampleSize = getSampleSize(encodedImage, rotationOptions, resizeOptions);
    final BitmapFactory.Options options = new BitmapFactory.Options();
    options.inSampleSize = sampleSize;
    Bitmap resizedBitmap;
    try {
      resizedBitmap = BitmapFactory.decodeStream(encodedImage.getInputStream(), null, options);
    } catch (OutOfMemoryError oom) {
      FLog.e(TAG, "Out-Of-Memory during transcode", oom);
      return new ImageTranscodeResult(TranscodeStatus.TRANSCODING_ERROR);
    }

    if (resizedBitmap == null) {
      FLog.e(TAG, "Couldn't decode the EncodedImage InputStream ! ");
      return new ImageTranscodeResult(TranscodeStatus.TRANSCODING_ERROR);
    }

    Matrix transformationMatrix =
        JpegTranscoderUtils.getTransformationMatrix(encodedImage, rotationOptions);

    Bitmap srcBitmap = resizedBitmap;
    try {
      if (transformationMatrix != null) {
        srcBitmap =
            Bitmap.createBitmap(
                resizedBitmap,
                0,
                0,
                resizedBitmap.getWidth(),
                resizedBitmap.getHeight(),
                transformationMatrix,
                false);
      }
      srcBitmap.compress(getOutputFormat(outputFormat), quality, outputStream);
      return new ImageTranscodeResult(
          sampleSize > DownsampleUtil.DEFAULT_SAMPLE_SIZE
              ? TranscodeStatus.TRANSCODING_SUCCESS
              : TranscodeStatus.TRANSCODING_NO_RESIZING);
    } catch (OutOfMemoryError oom) {
      FLog.e(TAG, "Out-Of-Memory during transcode", oom);
      return new ImageTranscodeResult(TranscodeStatus.TRANSCODING_ERROR);
    } finally {
      srcBitmap.recycle();
      resizedBitmap.recycle();
    }
  }

  @Override
  public boolean canResize(
      EncodedImage encodedImage,
      @Nullable RotationOptions rotationOptions,
      @Nullable ResizeOptions resizeOptions) {
    if (rotationOptions == null) {
      rotationOptions = RotationOptions.autoRotate();
    }
    return mResizingEnabled
        && DownsampleUtil.determineSampleSize(
                rotationOptions, resizeOptions, encodedImage, mMaxBitmapSize)
            > DownsampleUtil.DEFAULT_SAMPLE_SIZE;
  }

  @Override
  public boolean canTranscode(ImageFormat imageFormat) {
    return imageFormat == DefaultImageFormats.HEIF || imageFormat == DefaultImageFormats.JPEG;
  }

  @Override
  public String getIdentifier() {
    return "SimpleImageTranscoder";
  }

  private int getSampleSize(
      final EncodedImage encodedImage,
      final RotationOptions rotationOptions,
      @Nullable final ResizeOptions resizeOptions) {
    int sampleSize;
    if (!mResizingEnabled) {
      sampleSize = DownsampleUtil.DEFAULT_SAMPLE_SIZE;
    } else {
      sampleSize =
          DownsampleUtil.determineSampleSize(
              rotationOptions, resizeOptions, encodedImage, mMaxBitmapSize);
    }
    return sampleSize;
  }

  /**
   * Determine the {@link Bitmap.CompressFormat} given the {@link ImageFormat}. If no match is
   * found, it returns {@link Bitmap.CompressFormat#JPEG}
   *
   * @param format The {@link ImageFormat} used as input
   * @return The corresponding {@link Bitmap.CompressFormat}
   */
  private static Bitmap.CompressFormat getOutputFormat(@Nullable final ImageFormat format) {
    if (format == null) {
      return Bitmap.CompressFormat.JPEG;
    }

    if (format == DefaultImageFormats.JPEG) {
      return Bitmap.CompressFormat.JPEG;
    } else if (format == DefaultImageFormats.PNG) {
      return Bitmap.CompressFormat.PNG;
    } else {
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.ICE_CREAM_SANDWICH
          && DefaultImageFormats.isStaticWebpFormat(format)) {
        return Bitmap.CompressFormat.WEBP;
      }

      return Bitmap.CompressFormat.JPEG;
    }
  }
}

File: imagepipeline/src/main/java/com/facebook/imagepipeline/transcoder/SimpleImageTranscoder.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.imagepipeline.transcoder

import android.graphics.Bitmap
import android.graphics.Bitmap.CompressFormat
import android.graphics.BitmapFactory
import android.graphics.ColorSpace
import android.os.Build
import com.facebook.common.logging.FLog
import com.facebook.imageformat.DefaultImageFormats
import com.facebook.imageformat.ImageFormat
import com.facebook.imagepipeline.common.ResizeOptions
import com.facebook.imagepipeline.common.RotationOptions
import com.facebook.imagepipeline.image.EncodedImage
import java.io.OutputStream

/**
 * Image transcoder using only the Android API. Clients can use this if they don't want to use the
 * native implementation. This image transcoder requires more memory.
 */
class SimpleImageTranscoder(private val resizingEnabled: Boolean, private val maxBitmapSize: Int) :
    ImageTranscoder {

  override fun transcode(
      encodedImage: EncodedImage,
      outputStream: OutputStream,
      rotationOptions: RotationOptions?,
      resizeOptions: ResizeOptions?,
      outputFormat: ImageFormat?,
      quality: Int?,
      colorSpace: ColorSpace?,
  ): ImageTranscodeResult {
    var rotationOptions = rotationOptions
    var quality = quality
    if (quality == null) {
      quality = JpegTranscoderUtils.DEFAULT_JPEG_QUALITY
    }
    if (rotationOptions == null) {
      rotationOptions = RotationOptions.autoRotate()
    }
    val sampleSize = getSampleSize(encodedImage, rotationOptions, resizeOptions)
    val options = BitmapFactory.Options()
    options.inSampleSize = sampleSize
    if (colorSpace != null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      options.inPreferredColorSpace = colorSpace
    }
    val resizedBitmap: Bitmap? =
        try {
          BitmapFactory.decodeStream(encodedImage.inputStream, null, options)
        } catch (oom: OutOfMemoryError) {
          FLog.e(TAG, "Out-Of-Memory during transcode", oom)
          return ImageTranscodeResult(TranscodeStatus.TRANSCODING_ERROR)
        }
    if (resizedBitmap == null) {
      FLog.e(TAG, "Couldn't decode the EncodedImage InputStream ! ")
      return ImageTranscodeResult(TranscodeStatus.TRANSCODING_ERROR)
    }
    val transformationMatrix =
        JpegTranscoderUtils.getTransformationMatrix(encodedImage, rotationOptions)
    var srcBitmap: Bitmap = resizedBitmap
    return try {
      if (transformationMatrix != null) {
        srcBitmap =
            Bitmap.createBitmap(
                resizedBitmap,
                0,
                0,
                resizedBitmap.width,
                resizedBitmap.height,
                transformationMatrix,
                false)
      }
      srcBitmap.compress(getOutputFormat(outputFormat), quality, outputStream)
      ImageTranscodeResult(
          if (sampleSize > DownsampleUtil.DEFAULT_SAMPLE_SIZE) TranscodeStatus.TRANSCODING_SUCCESS
          else TranscodeStatus.TRANSCODING_NO_RESIZING)
    } catch (oom: OutOfMemoryError) {
      FLog.e(TAG, "Out-Of-Memory during transcode", oom)
      ImageTranscodeResult(TranscodeStatus.TRANSCODING_ERROR)
    } finally {
      srcBitmap.recycle()
      resizedBitmap.recycle()
    }
  }

  override fun canResize(
      encodedImage: EncodedImage,
      rotationOptions: RotationOptions?,
      resizeOptions: ResizeOptions?
  ): Boolean {
    var rotationOptions = rotationOptions
    if (rotationOptions == null) {
      rotationOptions = RotationOptions.autoRotate()
    }
    return resizingEnabled &&
        DownsampleUtil.determineSampleSize(
            rotationOptions, resizeOptions, encodedImage, maxBitmapSize) >
            DownsampleUtil.DEFAULT_SAMPLE_SIZE
  }

  override fun canTranscode(imageFormat: ImageFormat): Boolean =
      imageFormat === DefaultImageFormats.HEIF || imageFormat === DefaultImageFormats.JPEG

  override val identifier: String = "SimpleImageTranscoder"

  private fun getSampleSize(
      encodedImage: EncodedImage,
      rotationOptions: RotationOptions,
      resizeOptions: ResizeOptions?
  ): Int {
    val sampleSize: Int =
        if (!resizingEnabled) {
          DownsampleUtil.DEFAULT_SAMPLE_SIZE
        } else {
          DownsampleUtil.determineSampleSize(
              rotationOptions, resizeOptions, encodedImage, maxBitmapSize)
        }
    return sampleSize
  }

  companion object {
    private const val TAG = "SimpleImageTranscoder"

    /**
     * Determine the [Bitmap.CompressFormat] given the [ImageFormat]. If no match is found, it
     * returns [Bitmap.CompressFormat#JPEG]
     *
     * @param format The [ImageFormat] used as input
     * @return The corresponding [Bitmap.CompressFormat]
     */
    private fun getOutputFormat(format: ImageFormat?): CompressFormat {
      if (format == null) {
        return CompressFormat.JPEG
      }
      return if (format === DefaultImageFormats.JPEG) {
        CompressFormat.JPEG
      } else if (format === DefaultImageFormats.PNG) {
        CompressFormat.PNG
      } else {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.ICE_CREAM_SANDWICH &&
            DefaultImageFormats.isStaticWebpFormat(format)) {
          CompressFormat.WEBP
        } else {
          CompressFormat.JPEG
        }
      }
    }
  }
}
